source/particle/generator/probability_density_function.cc

```diff
@@ -0,0 +1,295 @@
+        const types::particle_index start_id = llround(n_tracers * start_weight / global_function_integral) + 1;
+
+        return generate_particles_in_subdomain(cell_weights,global_function_integral,start_weight,n_tracers, 1.1 * start_id);
+      }
+
+      template <int dim>
+      std::vector<double>
+      ProbabilityDensityFunction<dim>::local_accumulated_cell_weights () const
+      {
+        //evaluate function at all cell midpoints, sort cells according to weight
+        const QMidpoint<dim> quadrature_formula;
+        const unsigned int n_quadrature_points = quadrature_formula.size();
+
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_quadrature_points |
+                                 update_JxW_values);
+
+        std::vector<double> accumulated_cell_weights;
+
+        // compute the integral weight by quadrature
+        typename DoFHandler<dim>::active_cell_iterator
+        cell = this->get_dof_handler().begin_active(),
+        endc = this->get_dof_handler().end();
+        for (; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            {
+              fe_values.reinit (cell);
+              const std::vector<Point<dim> > position = fe_values.get_quadrature_points();
+
+              // Then add the weight of the current cell. Weights are always
+              // interpreted positively, even if the function evaluates to a
+              // negative number.
+              double next_cell_weight = 0.0;
+              if (accumulated_cell_weights.size() > 0)
+                next_cell_weight = accumulated_cell_weights.back();
+
+              for (unsigned int q = 0; q < n_quadrature_points; ++q)
+                next_cell_weight += std::fabs(function.value(position[q]) * fe_values.JxW(q));
+
+              // Start from the weight of the previous cell
+              accumulated_cell_weights.push_back(next_cell_weight);
+            }
+        return accumulated_cell_weights;
+      }
+
+      template <int dim>
+      std::multimap<types::LevelInd, Particle<dim> >
+      ProbabilityDensityFunction<dim>::generate_particles_in_subdomain (const std::map<double,types::LevelInd> &cells,
+                                                                        const double global_weight,
+                                                                        const double start_weight,
+                                                                        const types::particle_index num_particles,
+                                                                        const types::particle_index start_id)
+      {
+        std::multimap<types::LevelInd, Particle<dim> > particles;
+
+        // Pick cells and assign particles at random points inside them
+        for (types::particle_index cur_id = start_id; cur_id < start_id + num_particles; ++cur_id)
```

Comment:
Or maybe I misinterpret what `num_particles` corresponds to. Is this the *global* number of particles? If so, then it seems odd to run from `start_id` to `start_id ... num_particles`.